src/bootstrap/src/bin/rustc.rs

```diff
@@ -145,6 +145,35 @@ fn main() {
         cmd.arg("-Z").arg("on-broken-pipe=kill");
     }
 
+    // In `cargo.rs` we add the `FLAGS_FOR_RUSTC` flags for all `compiler/`
+    // crates to RUSTFLAGS. However, RUSTFLAGS is ignored for `compiler/` proc
+    // macro crates, because RUSTFLAGS is ignored when `--target` is given. So,
+    // we add the `FLAGS_FOR_RUSTC` flags directly to the command here if they
+    // are also in RUSTFLAGS.
+    let crate_type = parse_value_from_args(&orig_args, "--crate-type");
+    if crate_type == Some("proc-macro") {
+        if let Ok(rustflags) = env::var("RUSTFLAGS") {
+            // Count how many of flags from `FLAGS_FOR_RUSTC` are in RUSTFLAGS.
```

Comment:
Why do we have to use a fragile hacky heuristic like that? That shouldn't be needed. Bootstrap should clearly tell the wrapper what to add where, we should never do guesswork like this.

Reply:
I know very little about bootstrap and this was something I managed to get working. I'm happy to hear other ideas and implement them but I'll need a more detailed explanation than "bootstrap should clearly tell the wrapper" - how and where would this be done? :)

Reply:
I was imagining something like this, which already exists:

https://github.com/rust-lang/rust/blob/c85d3f6e874acd35647617161fd2ae950b94dedd/src/bootstrap/src/bin/rustc.rs#L138-L140

From some of the other discussion it sounds like what you dislike about this approach is that it will pass the flags to all invocations, not just the ones for workspace crates. However, non-workspace crates run with `--cap-lints` so that should be entirely non-consequential?

The discussion about this seems to be spread across multiple Zulip topics, issues, and PRs, so I have no idea which options have been considered. Has `RUSTC_LINT_FLAGS` been discarded for some reason?

Reply:
[This thread](https://rust-lang.zulipchat.com/#narrow/channel/233931-t-compiler.2Fmajor-changes/topic/Enable.20.60warn.28unused_crate_dependencies.29.20o.E2.80.A6.20compiler-team.23844) has most of the discussion.

I didn't even know RUSTC_LINT_FLAGS was a thing until just now. Like I said above (and in the thread), I know very little about bootstrap and I've tried several different approaches based on suggestions people have made at different times.

Reply:
That's fair -- I just thought bootstrap people were involved in the discussion / review so I'm a bit surprised it took so long for someone to suggest this. Let's make sure to include bootstrap people in the next PR.

Reply:
I guess I am indeed underestimating how fragmented bootstrap knowledge is.

Reply:
I also had no idea about the flag :) bootstrap's build logic is distributed amongst several places, it's full of hacks and edge cases, and hidden invariants that can break when you modify completely unrelated code.

Reply:
As I said, I didn't know about the flag either. I was about to suggest to add something like this flag, went looking to the sources for inspiration on how to call it, and then saw that it already existed.

Reply:
Same 😆